codegen/test_data.py

```diff
@@ -23,72 +23,101 @@ def test_parse_can_message_enum_invalid_can_id(self, mock_open):  # pylint: disa
-            data.parse_can_message_enum(None)
-
-    @mock.patch("data.open")
-    def test_parse_can_message_enum_duplicate_can_id(self, mock_open):  # pylint: disable=invalid-name
-        """Tests there is a check for a duplicate CAN ID."""
-        ascii_protobuf = """
+        """), ("""
             msg {
-                id:1
+                id: 1
                 msg_name: "can message 1"
                 can_data {
-                    frame {
-                        type: UINT64
+                    empty {
                     }
                 }
             }
 
             msg {
-                id:1
+                id: 1
                 msg_name: "silly duplicate message id"
                 can_data {
-                    frame {
-                        type: UINT64
+                    empty {
+                    }
+                }
+            }
+        """), ("""
+            msg {
+                id: 1
+                msg_name: "some name"
+                can_data {
+                    empty {
+                    }
+                }
+            }
+
+            msg {
+                id: 3
+                msg_name: "some name"
+                can_data {
+                    empty {
                     }
                 }
             }
-        """
+        """)])
+    @mock.patch("data.open")
+    def test_parse_can_message_enum_exceptions(self, ascii_protobuf,
+                                               mock_open):
+        """Tests for an invalid CAN ID handler."""
         mock_open.side_effect = [
             mock.mock_open(read_data=ascii_protobuf).return_value
         ]
         with self.assertRaises(Exception):
             data.parse_can_message_enum(None)
 
-    @mock.patch("data.open")
-    def test_parse_can_message_enum_duplicate_name(self, mock_open):  # pylint: disable=invalid-name
-        """Checks for a duplicate CAN message name."""
-        ascii_protobuf = """
+    @parameterized.expand([("""
             msg {
-                id:1
-                msg_name: "some name"
+                id:300
+                msg_name: "can id over 128"
                 can_data {
                     frame {
                         type: UINT64
                     }
                 }
             }
+        """), ("""
+            msg {
+                id: 1
+                msg_name: "can message 1"
+                can_data {
+                    empty {
+                    }
+                }
+            }
 
             msg {
-                id:3
+                id: 1
+                msg_name: "silly duplicate message id"
+                can_data {
+                    empty {
+                    }
+                }
+            }
+        """), ("""
+            msg {
+                id: 1
                 msg_name: "some name"
                 can_data {
-                    frame {
-                        type: UINT64
+                    u8 {
```

Comment:
If you're declaring fields like this, wouldn't it be pretty easy to count the number of fields you're using?

Reply:
Yes? What are you trying to get at though?

I already use that fact to mask the fields.  I'm now planning to use that for the DLC as well. The reason it isn't a repeated field is so that codegen doesn't have to enforce the number of fields in the gen process which makes it more robust.

Reply:
LOL sorry this was for the DLC